src/Types/BlockItemSummary.cs

```diff
@@ -0,0 +1,368 @@
+using Concordium.Grpc.V2;
+using Concordium.Sdk.Exceptions;
+
+namespace Concordium.Sdk.Types;
+
+/// <summary>
+/// Summary of the outcome of a block item in structured form.
+/// The summary determines which transaction type it was.
+/// </summary>
+public sealed class BlockItemSummary
+{
+    private readonly Grpc.V2.BlockItemSummary _blockItemSummary;
+
+    internal BlockItemSummary(Grpc.V2.BlockItemSummary blockItemSummary) => this._blockItemSummary = blockItemSummary;
+
+
+    /// <summary>
+    /// Return whether the transaction was successful, i.e., the intended effect
+    /// happened.
+    /// </summary>
+    /// <returns>bool representing if transaction succeeded.</returns>
+    /// <exception cref="MissingEnumException{DetailsOneofCase}">Throws exception when returned type not known</exception>
+    public bool IsSuccess() =>
+        this._blockItemSummary.DetailsCase switch
+        {
+            Grpc.V2.BlockItemSummary.DetailsOneofCase.AccountTransaction =>
+                !AccountTransactionDetailsIsRejected(this._blockItemSummary.AccountTransaction, out _),
+            Grpc.V2.BlockItemSummary.DetailsOneofCase.AccountCreation => true,
+            Grpc.V2.BlockItemSummary.DetailsOneofCase.Update => true,
+            Grpc.V2.BlockItemSummary.DetailsOneofCase.None =>
+                throw new MissingEnumException<Grpc.V2.BlockItemSummary.DetailsOneofCase>(Grpc.V2.BlockItemSummary.DetailsOneofCase.None),
+            _ =>
+                throw new MissingEnumException<Grpc.V2.BlockItemSummary.DetailsOneofCase>(this._blockItemSummary.DetailsCase)
+        };
+
+    /// <summary>
+    /// Return whether the transaction has failed to achieve the intended
+    /// effects.
+    /// </summary>
+    /// <returns>bool representing if transaction is rejected</returns>
+    public bool IsReject() => this.TryGetRejectedAccountTransaction(out _);
+
+    /// <summary>
+    /// When transaction is rejected this method will return true and rejected reason
+    /// will be return in out param.
+    /// </summary>
+    /// <param name="rejectReason">Output with rejected reason if rejected.</param>
+    /// <returns>bool representing if transaction is rejected.</returns>
+    /// <exception cref="MissingEnumException{DetailsOneofCase}">Throws exception when returned type not known</exception>
+    public bool TryGetRejectedAccountTransaction(out RejectReason? rejectReason)
```

Comment:
Is this the way to do things in C#? I would have thought returning an Option type would be the way to go, rather than out parameters.

Reply:
Yes - there are some language extension packages which includes options and those functional language operators but it isn’t very C# alike.

I wrapped my head around this a few times when I wrote the code and I’m still convinced this is the best C# approach. Try… patterns is well known in the C# world.

Reply:
I have had a talk with Ali and this is the best approach as we see it.